src/logql-query.ts

```diff
@@ -1,48 +1,250 @@
+      }
+    },
+  });
+
+  return pipelineStages;
+};
 
 export class LogQLQuery {
-  selectors: Array<string> = [];
-  pipeline: Array<string> = [];
+  streamSelector: Array<LabelMatcher> = [];
+  pipeline: Array<PipelineStage> = [];
 
   constructor(query: string) {
-    const parts = query.match(/.*{([^}]*)}(.*)/);
-    this.selectors =
-      parts && parts[1].length > 0
-        ? parts[1].split(',').filter(notEmptyString).map(trim)
-        : [];
-    this.pipeline =
-      parts && parts[2].length > 0
-        ? parts[2].split('|').filter(notEmptyString).map(trim)
-        : [];
+    const syntaxTree = parser.parse(query);
+    let parsedMatchers: Array<LabelMatcher> = [];
+    let parsedPipeline: Array<PipelineStage> = [];
+
+    syntaxTree.iterate({
+      enter(node) {
+        if (node.name) {
+          if (node.name === 'Selector' && parsedMatchers.length === 0) {
+            parsedMatchers = parseMatchers(syntaxTree, node, query);
+
+            return false;
+          } else if (
+            node.name === 'PipelineExpr' &&
+            parsedPipeline.length === 0
+          ) {
+            parsedPipeline = parsePipelineStages(syntaxTree, node, query);
+
+            return false;
+          }
+        }
+      },
+    });
+
+    this.streamSelector = parsedMatchers;
+    this.pipeline = parsedPipeline;
   }
 
-  appendSelector = (newSelectors: string | Array<string>) => {
-    if (Array.isArray(newSelectors)) {
-      this.selectors.push(...newSelectors);
-    } else {
-      this.selectors.push(newSelectors);
+  removeSelectorMatcher = (matcherToRemove: Partial<LabelMatcher>) => {
+    this.streamSelector = this.streamSelector.filter((matcher) => {
+      let include = true;
+
+      if (matcherToRemove.label) {
+        include = include && matcher.label !== matcherToRemove.label;
+      }
+
+      return include;
+    });
+
+    return this;
+  };
+
+  addSelectorMatcher = (matchers?: LabelMatcher | Array<LabelMatcher>) => {
+    if (!matchers) {
+      return this;
     }
 
-    this.selectors = this.selectors.filter(notEmptyString).map(trim);
+    const matchersArray = Array.isArray(matchers) ? matchers : [matchers];
+
+    matchersArray.forEach((matcher) => {
+      const existingSelectorIndex = this.streamSelector.findIndex(
+        ({ label }) => matcher.label === label,
+      );
+
+      if (existingSelectorIndex !== -1) {
+        this.streamSelector[existingSelectorIndex] = matcher;
+      } else {
+        this.streamSelector.push(matcher);
+      }
+    });
+
+    this.streamSelector = this.streamSelector.filter(
+      (selector, index) =>
+        this.streamSelector.findIndex(
+          ({ label }) => selector.label === label,
+        ) === index,
+    );
+
+    return this;
+  };
+
+  removePipelineStage = (
+    pipelineStageToRemove: Partial<PipelineStage>,
+    matchOptions?: { matchLabel?: string },
+  ) => {
+    this.pipeline = this.pipeline.filter((existingStage) => {
```

Comment:
What should be the behavior if one part of `pipelineStageToRemove` matches, but not all? (e.g. `operator` matches, but `value` doesn't). Maybe worth adding a test or two for this case.

Reply:
any match will remove a stage, I added some tests.

Reply:
Probably a bit easier using early returns instead of `include` (not tested):

```
this.pipeline = this.pipeline.filter((existingStage) => {
  if (pipelineStageToRemove.operator &&
    existingStage.operator === pipelineStageToRemove.operator) {
    return false;
  }

  if (pipelineStageToRemove.value &&
    existingStage.value === pipelineStageToRemove.value) {
    return false;
  }

  if (matchOptions?.matchLabel &&
    existingStage.labelsInFilter?.every((label) => label.label === matchOptions.matchLabel) {
    return false;
  }

  return true;
});
```